File: platform/robot/src/hardware/motors/servo/driver.py
"""RC servo steering driver (single servo, counter-phase four-wheel steering).

One servo drives both axles through the linkage: the rear wheels turn by the
same amount as the front but in the OPPOSITE direction (confirmed on hardware
2026-07-25 -- this was previously documented as "parallel/crab steering",
which is the opposite geometry). The consequence for anything reasoning about
motion is that the chassis pivots about its centre, not the rear axle, and
yaws twice as fast as a front-steer car at the same angle -- see
src/simulation/kinematics.py.

Implements the ``SteeringDriver`` port for a hobby RC servo driven by a 50 Hz
PWM signal, using the kernel's **hardware** PWM peripheral via
``/sys/class/pwm``. The servo has no position feedback: reported speed is
always 0 and reported position is the last commanded value.

Why sysfs and not gpiozero: gpiozero's ``PWMOutputDevice`` under the pin
factory the Pi Zero actually uses (``LGPIOFactory``) generates the pulse train
in *software*, so kernel scheduling jitter lands directly on the servo's pulse
width and the servo visibly twitches even while holding a fixed angle.
Measured on hardware 2026-07-25: the twitching persisted with a single command
and no PWM rewrites at all, and on a fresh battery, which ruled out both
command traffic and supply sag. The hardware peripheral is immune to CPU load.

This requires the PWM overlay to be mapped onto the servo's pin. On this
robot (servo on GPIO 12) that is::

    dtoverlay=pwm,pin=12,func=4      # func=4 is ALT0, GPIO 12's PWM function

in ``/boot/firmware/config.txt`` (applied by ``scripts/provisioning/bootstrap-fresh-zero.sh``),
followed by a reboot. Without it there is no ``pwmchip`` to open and
:meth:`connect` fails with a message saying so -- deliberately, rather than
silently falling back to jittery software PWM.

Access is group-based, not root-only: Raspberry Pi OS ships a udev rule
(``99-com.rules``) that chgrp's ``/sys/class/pwm`` to ``gpio``, and the service
user is in that group.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING, override

from src.hardware.exceptions import MotorConnectionError
from src.hardware.motors.base import (
    DEFAULT_STEERING_SPEED,
    STEERING_CENTER_DEG,
    SteeringDriver,
)
from src.hardware.motors.pwm_sysfs import (
    EXPORT_TIMEOUT_S,
    SYSFS_PWM_ROOT,
    wait_for_pwm_channel_writable,
)
from src.hardware.motors.servo.config import (
    NS_PER_US,
    PWM_FREQUENCY_HZ,
    US_PER_SECOND,
    ServoConfig,
)

if TYPE_CHECKING:
    from pathlib import Path

logger = logging.getLogger(__name__)

_PERIOD_NS = int(US_PER_SECOND / PWM_FREQUENCY_HZ) * NS_PER_US
"""One PWM frame in nanoseconds (20 ms at 50 Hz)."""

_PULSE_EPSILON_US = 1.0
"""Smallest pulse-width change worth writing.

``move_steering_to`` is called once per incoming ``/ackermann_cmd``, and a
controller holding a steady heading re-sends the same angle indefinitely.
Skipping identical writes avoids pointless syscalls. 1 us is well under a
hobby servo's own deadband (typically 2-10 us), so this cannot swallow a
movement the servo would have acted on.
"""


class Driver(SteeringDriver):
    """Single-servo steering via the kernel's hardware PWM (50 Hz, sysfs)."""

    def __init__(self, config: ServoConfig | None = None) -> None:
        """Build the driver without touching hardware (see :meth:`connect`).

        Args:
            config: Servo configuration; defaults to env-derived ``ServoConfig``.
        """
        self._config = config or ServoConfig()
        self._position = STEERING_CENTER_DEG
        self._channel_dir: Path | None = None
        self._pulse_us: float | None = None

    @property
    def _chip_dir(self) -> Path:
        return SYSFS_PWM_ROOT / f"pwmchip{self._config.pwmchip}"

    def _fail(self, reason: str) -> MotorConnectionError:
        return MotorConnectionError(str(self._config.gpio_pin), reason)

    def _export_channel(self) -> Path:
        """Export the PWM channel and wait for it to become writable."""
        chip = self._chip_dir
        if not chip.is_dir():
            msg = (
                f"{chip} not present -- hardware PWM overlay missing. Add "
                f"'dtoverlay=pwm,pin={self._config.gpio_pin},func=4' to "
                "/boot/firmware/config.txt and reboot"
            )
            raise self._fail(
                msg,
            )

        channel_dir = chip / f"pwm{self._config.pwm_channel}"
        if not channel_dir.is_dir():
            try:
                (chip / "export").write_text(str(self._config.pwm_channel))
            except OSError as err:
                # EBUSY means someone already exported it, which is fine.
                if not channel_dir.is_dir():
                    msg = f"cannot export PWM channel: {err}"
                    raise self._fail(msg) from err

        if wait_for_pwm_channel_writable(channel_dir):
            return channel_dir

        msg = (
            f"{channel_dir} did not become writable within {EXPORT_TIMEOUT_S}s "
            "(is the service user in the 'gpio' group?)"
        )
        raise self._fail(
            msg,
        )

    @override
    def connect(self) -> None:
        """Export the PWM channel, start the 50 Hz carrier, and centre the wheels.

        Raises:
            MotorConnectionError: overlay missing, or the channel never became
                writable.
        """
        channel_dir = self._export_channel()

        try:
            # Order matters: duty_cycle may never exceed period, so a stale
            # larger duty from a previous run would make the period write
            # fail. Zero the duty first, then set the frame, then enable.
            (channel_dir / "duty_cycle").write_text("0")
            (channel_dir / "period").write_text(str(_PERIOD_NS))
            (channel_dir / "enable").write_text("1")
        except OSError as err:
            msg = f"PWM init failed: {err}"
            raise self._fail(msg) from err

        self._channel_dir = channel_dir
        self.center_steering()
        logger.info(
            "Servo steering connected on hardware PWM %s (GPIO %d)",
            channel_dir,
            self._config.gpio_pin,
        )

    @override
    def disconnect(self) -> None:
        """Stop the PWM carrier and release the channel."""
        if self._channel_dir is not None:
            try:
                (self._channel_dir / "enable").write_text("0")
            except OSError:
                logger.warning("Failed to disable servo PWM on disconnect", exc_info=True)
            self._channel_dir = None
        # Forget the cached pulse so the first write after reconnecting is not
        # skipped as a no-op (the carrier was disabled in between).
        self._pulse_us = None

    def _position_to_pulse_us(self, position_deg: float) -> float:
        """Map an absolute steering angle (deg, 0 = centre) to a pulse width (us)."""
        cfg = self._config
        signed = -position_deg if cfg.reversed else position_deg
        span_us = cfg.max_pulse_us - cfg.min_pulse_us
        pulse_us = cfg.center_pulse_us + (signed / cfg.range_deg) * span_us
        return max(cfg.min_pulse_us, min(cfg.max_pulse_us, pulse_us))

    @override
    def move_steering_to(self, position: float, speed: int = DEFAULT_STEERING_SPEED) -> None:
        """Move steering to an absolute angle in degrees.

        ``speed`` is accepted for interface parity but ignored: an RC servo
        self-paces to its commanded position with no host-side rate control.

        Raises:
            MotorConnectionError: if called before :meth:`connect`, or the
                write fails.
        """
        del speed  # servo self-paces; no host-side rate control
        if self._channel_dir is None:
            msg = "driver not connected"
            raise self._fail(msg)

        pulse_us = self._position_to_pulse_us(position)
        if self._pulse_us is None or abs(pulse_us - self._pulse_us) >= _PULSE_EPSILON_US:
            try:
                (self._channel_dir / "duty_cycle").write_text(str(int(pulse_us * NS_PER_US)))
            except OSError as err:
                msg = f"PWM duty_cycle write failed: {err}"
                raise self._fail(msg) from err
            self._pulse_us = pulse_us
        self._position = position

    @override
    def get_steering_position(self) -> float:
        """Last commanded steering angle (servo has no position feedback)."""
        return self._position

File: src/go/pkg/driver/servo/driver.go
package servo

import (
	"context"
	"errors"
	"fmt"
	"math"

	"github.com/go-playground/validator/v10"

	"github.com/teamvoltimor/vtitan/src/go/pkg/driver/internal/sysfspwm"
)

// Config is the servo's wiring and pulse calibration. Every field except
// Root mirrors a key of src/config/hardware/motors/servo.toml (and the
// Python ServoConfig built from it); there is deliberately no DefaultConfig,
// because the active servo profile overlays range_deg (a 270 degree servo
// read with the 180 degree base value would steer at two thirds of the
// commanded angle). Resolve it from the TOML tree instead.
type Config struct {
	// Root is the sysfs PWM class directory; empty means
	// sysfspwm.DefaultRoot (/sys/class/pwm). Tests point it at a temp dir.
	Root string

	// GPIOPin is only used to name the overlay in Export's error.
	GPIOPin    int `validate:"gte=0"`
	PWMChip    int `validate:"gte=0"`
	PWMChannel int `validate:"gte=0"`
	// FrequencyHz is the carrier; 50 Hz for a hobby servo.
	FrequencyHz int `validate:"required,gt=0"`

	MinPulseUS    float64 `validate:"gt=0"`
	MaxPulseUS    float64 `validate:"gtfield=MinPulseUS"`
	CenterPulseUS float64 `validate:"gtefield=MinPulseUS,ltefield=MaxPulseUS"`
	// RangeDeg is the servo's full mechanical travel (180 or 270): the angle
	// that spans MinPulseUS..MaxPulseUS.
	RangeDeg float64 `validate:"gt=0"`
	// Reversed flips the sign of every commanded angle.
	Reversed bool
}

// channel is the part of sysfspwm.Channel the driver uses once exported.
// It is an interface so the connect/center/epsilon/close sequencing can be
// tested with a recorder.
type channel interface {
	Init() error
	WriteDutyNS(dutyNS int64) error
	Disable() error
}

// Driver is a single RC servo on a hardware PWM channel. It has no position
// feedback: Angle is the last commanded value. Not safe for concurrent use;
// the motor loop owns it from one goroutine.
type Driver struct {
	cfg Config

	ch       channel
	angleDeg float64
	// pulseUS is the last pulse written; hasPulse is false before the first
	// write after Connect, so that write is never skipped as a no-op
	// (servo/driver.py:89 _pulse_us = None, reset on disconnect at :147).
	pulseUS  float64
	hasPulse bool
}

const (
	// CenterDeg is the servo angle for wheels-straight
	// (motors/base.py:40 STEERING_CENTER_DEG). Center writes it; the
	// steering offset trim is NOT applied, as in Python's center_steering.
	CenterDeg = 0.0

	// pulseEpsilonUS is the smallest pulse change worth writing
	// (servo/driver.py:63 _PULSE_EPSILON_US). A controller holding a
	// heading resends the same angle every tick; 1 us is well under a hobby
	// servo's own 2-10 us deadband, so this cannot swallow a real move.
	pulseEpsilonUS = 1.0

	// nsPerUS converts the pulse width to the duty_cycle attribute's unit.
	nsPerUS = 1000
)

var (
	// ErrNotConnected is returned by SetAngle and Center before Connect.
	ErrNotConnected = errors.New("servo: driver not connected")

	// ErrNonFiniteAngle is returned by SetAngle for NaN or +-Inf, and
	// nothing is written: the pulse stays where it was. Python has no such
	// guard - NaN falls through its min/max clamp to a full-lock pulse - so
	// this is a deliberate divergence, not a port.
	ErrNonFiniteAngle = errors.New("servo: non-finite angle")
)

// New validates cfg and returns a Driver. Call Connect before SetAngle.
func New(cfg Config) (*Driver, error) {
	if err := validator.New().Struct(cfg); err != nil {
		return nil, fmt.Errorf("servo: invalid config: %w", err)
	}
	if cfg.Root == "" {
		cfg.Root = sysfspwm.DefaultRoot
	}
	return &Driver{cfg: cfg, angleDeg: CenterDeg}, nil
}

// Connect exports the PWM channel, starts the carrier (duty 0, period,
// enable, in that order), and centers the servo - servo/driver.py:108-134.
// On a machine without the PWM overlay (any dev box) it fails with an error
// wrapping sysfspwm.ErrOverlayMissing.
func (d *Driver) Connect(ctx context.Context) error {
	ch := sysfspwm.New(
		d.cfg.Root,
		d.cfg.PWMChip,
		d.cfg.PWMChannel,
		d.cfg.FrequencyHz,
		// servo/driver.py:103's hint, verbatim.
		fmt.Sprintf("'dtoverlay=pwm,pin=%d,func=4'", d.cfg.GPIOPin),
	)
	if err := ch.Export(ctx); err != nil {
		return fmt.Errorf("servo: GPIO %d: %w", d.cfg.GPIOPin, err)
	}
	return d.connectChannel(ch)
}

// SetAngle moves the servo to an absolute angle in servo degrees, 0 =
// center, positive = the direction a positive pulse offset turns it (after
// Reversed). The pulse is clamped to [MinPulseUS, MaxPulseUS]; a change
// under 1 us is not written. servo/driver.py:157-181 move_steering_to.
func (d *Driver) SetAngle(angleDeg float64) error {
	if math.IsNaN(angleDeg) || math.IsInf(angleDeg, 0) {
		return fmt.Errorf("%w: %v", ErrNonFiniteAngle, angleDeg)
	}
	if d.ch == nil {
		return ErrNotConnected
	}

	pulseUS := PulseUS(d.cfg, angleDeg)
	if !d.hasPulse || math.Abs(pulseUS-d.pulseUS) >= pulseEpsilonUS {
		// int(pulse_us * NS_PER_US): truncation toward zero, as Python's int().
		if err := d.ch.WriteDutyNS(int64(pulseUS * nsPerUS)); err != nil {
			return fmt.Errorf("servo: PWM duty_cycle write failed: %w", err)
		}
		d.pulseUS = pulseUS
		d.hasPulse = true
	}
	d.angleDeg = angleDeg
	return nil
}

// Center moves the servo to CenterDeg (motors/base.py:125-127).
func (d *Driver) Center() error {
	return d.SetAngle(CenterDeg)
}

// Angle is the last commanded servo angle in degrees (no feedback).
func (d *Driver) Angle() float64 { return d.angleDeg }

// Close disables the carrier and forgets the cached pulse, so the first
// write after a reconnect is not skipped (servo/driver.py:136-147). It does
// not center first, as Python's disconnect does not: callers that want the
// wheels straight center before closing, as the motor loop does on its way
// out. With the carrier off the servo gets no pulses at all; whether it then
// goes limp or holds its last position depends on the servo and has not been
// checked on this one. Close on an unconnected Driver is a no-op.
func (d *Driver) Close() error {
	d.hasPulse = false
	if d.ch == nil {
		return nil
	}
	ch := d.ch
	d.ch = nil
	if err := ch.Disable(); err != nil {
		return fmt.Errorf("servo: disabling PWM on close: %w", err)
	}
	return nil
}

// connectChannel is Connect after export: init the carrier, then center.
func (d *Driver) connectChannel(ch channel) error {
	if err := ch.Init(); err != nil {
		return fmt.Errorf("servo: PWM init failed: %w", err)
	}
	d.ch = ch
	d.hasPulse = false
	return d.Center()
}

// PulseUS maps a servo angle (deg, 0 = center) to a pulse width in
// microseconds: center + (angle / range) * (max - min), sign flipped when
// Reversed, clamped to [min, max]. servo/driver.py:149-155
// _position_to_pulse_us. The caller keeps angleDeg finite.
func PulseUS(cfg Config, angleDeg float64) float64 {
	signed := angleDeg
	if cfg.Reversed {
		signed = -angleDeg
	}
	spanUS := cfg.MaxPulseUS - cfg.MinPulseUS
	pulseUS := cfg.CenterPulseUS + (signed/cfg.RangeDeg)*spanUS
	return max(cfg.MinPulseUS, min(cfg.MaxPulseUS, pulseUS))
}
